### DISCUSSION TITLE
application getting exit whenever i am trying to sending notification in react native project

### DISCUSSION THREAD

**Discussion by ashish-hexalitics** [invertase/react-native-firebase#7510] [Q&A]
application getting exit whenever i am trying to sending notification in react native project
 react-native-firebase/messaging
 
 i am using this package   "@react-native-firebase/messaging": "^18.7.3"
 
 // Top-level build file where you can add configuration options common to all sub-projects/modules.

./android/build.gradle
`buildscript {
    ext {
        firebaseMessagingVersion:"21.1.0"
        buildToolsVersion = "33.0.0"
        minSdkVersion = 21
        compileSdkVersion = 33
        targetSdkVersion = 33

        // We use NDK 23 which hasm both M1 support and is the side-by-side NDK version from AGP.
        ndkVersion = "23.1.7779620"
    }
    repositories {
        google()
        mavenCentral()
    }

    dependencies {
        classpath("com.android.tools.build:gradle")
        classpath("com.facebook.react:react-native-gradle-plugin")
        classpath('com.google.gms:google-services:4.4.0')
    }
}`

./android/app/build.gradle

`apply plugin: "com.android.application"
apply plugin: "com.facebook.react"
apply plugin: 'com.google.gms.google-services' // <- Add this line

/**
 * This is the configuration block to customize your React Native Android app.
 * By default you don't need to apply any configuration, just uncomment the lines you need.
 */
react {
    /* Folders */
    //   The root of your project, i.e. where "package.json" lives. Default is '..'
    // root = file("../")
    //   The folder where the react-native NPM package is. Default is ../node_modules/react-native
    // reactNativeDir = file("../node_modules/react-native")
    //   The folder where the react-native Codegen package is. Default is ../node_modules/@react-native/codegen
    // codegenDir = file("../node_modules/@react-native/codegen")
    //   The cli.js file which is the React Native CLI entrypoint. Default is ../node_modules/react-native/cli.js
    // cliFile = file("../node_modules/react-native/cli.js")

    /* Variants */
    //   The list of variants to that are debuggable. For those we're going to
    //   skip the bundling of the JS bundle and the assets. By default is just 'debug'.
    //   If you add flavors like lite, prod, etc. you'll have to list your debuggableVariants.
    // debuggableVariants = ["liteDebug", "prodDebug"]

    /* Bundling */
    //   A list containing the node command and its flags. Default is just 'node'.
    // nodeExecutableAndArgs = ["node"]
    //
    //   The command to run when bundling. By default is 'bundle'
    // bundleCommand = "ram-bundle"
    //
    //   The path to the CLI configuration file. Default is empty.
    // bundleConfig = file(../rn-cli.config.js)
    //
    //   The name of the generated asset file containing your JS bundle
    // bundleAssetName = "MyApplication.android.bundle"
    //
    //   The entry file for bundle generation. Default is 'index.android.js' or 'index.js'
    // entryFile = file("../js/MyApplication.android.js")
    //
    //   A list of extra flags to pass to the 'bundle' commands.
    //   See https://github.com/react-native-community/cli/blob/main/docs/commands.md#bundle
    // extraPackagerArgs = []

    /* Hermes Commands */
    //   The hermes compiler command to run. By default it is 'hermesc'
    // hermesCommand = "$rootDir/my-custom-hermesc/bin/hermesc"
    //
    //   The list of flags to pass to the Hermes compiler. By default is "-O", "-output-source-map"
    // hermesFlags = ["-O", "-output-source-map"]
}

/**
 * Set this to true to Run Proguard on Release builds to minify the Java bytecode.
 */
def enableProguardInReleaseBuilds = false

/**
 * The preferred build flavor of JavaScriptCore (JSC)
 *
 * For example, to use the international variant, you can use:
 * `def jscFlavor = 'org.webkit:android-jsc-intl:+'`
 *
 * The international variant includes ICU i18n library and necessary data
 * allowing to use e.g. `Date.toLocaleString` and `String.localeCompare` that
 * give correct results when using with locales other than en-US. Note that
 * this variant is about 6MiB larger per architecture than default.
 */
def jscFlavor = 'org.webkit:android-jsc:+'

android {
    ndkVersion rootProject.ext.ndkVersion

    compileSdkVersion rootProject.ext.compileSdkVersion

    namespace "com.digianafresh"
    defaultConfig {
        applicationId "com.digianafresh"
        minSdkVersion rootProject.ext.minSdkVersion
        targetSdkVersion rootProject.ext.targetSdkVersion
        versionCode 37
        versionName "3.1.7"
    }
    signingConfigs {
        debug {
            storeFile file('debug.keystore')
            storePassword 'android'
            keyAlias 'androiddebugkey'
            keyPassword 'android'
        }
    }
    buildTypes {
        debug {
            signingConfig signingConfigs.debug
        }
        release {
            // Caution! In production, you need to generate your own keystore file.
            // see https://reactnative.dev/docs/signed-apk-android.
            signingConfig signingConfigs.debug
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
        }
    }
}

dependencies {
    // The version of react-native is set by the React Native Gradle Plugin
    implementation("com.facebook.react:react-android")
    debugImplementation("com.facebook.flipper:flipper:${FLIPPER_VERSION}")
    debugImplementation("com.facebook.flipper:flipper-network-plugin:${FLIPPER_VERSION}") {
        exclude group:'com.squareup.okhttp3', module:'okhttp'
    }

    debugImplementation("com.facebook.flipper:flipper-fresco-plugin:${FLIPPER_VERSION}")
    if (hermesEnabled.toBoolean()) {
        implementation("com.facebook.react:hermes-android")
    } else {
        implementation jscFlavor
    }
    implementation project(':payu-non-seam-less-react')
    implementation platform('com.google.firebase:firebase-bom:32.7.0')
    implementation('com.google.firebase:firebase-analytics')
    implementation('com.google.firebase:firebase-messaging')
}

apply from: "../../node_modules/react-native-vector-icons/fonts.gradle"
apply from: file("../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesAppBuildGradle(project)`

MainApplication.java

`package com.digianafresh;

import android.app.Application;
import com.facebook.react.PackageList;
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
import com.facebook.react.defaults.DefaultNewArchitectureEntryPoint;
import com.facebook.react.defaults.DefaultReactNativeHost;
import com.facebook.soloader.SoLoader;
import io.invertase.firebase.app.ReactNativeFirebaseAppPackage;
// import com.dieam.reactnativepushnotification.ReactNativePushNotificationPackage;
import com.google.firebase.messaging.FirebaseMessaging;
import java.util.List;

public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost = new DefaultReactNativeHost(this) {

    @Override
    public boolean getUseDeveloperSupport() {
      return BuildConfig.DEBUG;
    }

    @Override
    protected List<ReactPackage> getPackages() {
      @SuppressWarnings("UnnecessaryLocalVariable")
      List<ReactPackage> packages = new PackageList(this).getPackages();
      // Packages that cannot be autolinked yet can be added manually here, for
      // example:
      // packages.add(new MyReactNativePackage());
      return packages;
    }

    @Override
    protected String getJSMainModuleName() {
      return "index";
    }

    @Override
    protected boolean isNewArchEnabled() {
      return BuildConfig.IS_NEW_ARCHITECTURE_ENABLED;
    }

    @Override
    protected Boolean isHermesEnabled() {
      return BuildConfig.IS_HERMES_ENABLED;
    }
  };

  @Override
  public ReactNativeHost getReactNativeHost() {
    return mReactNativeHost;
  }

  @Override
  public void onCreate() {
    super.onCreate();
    SoLoader.init(this, /* native exopackage */ false);
    if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
      // If you opted-in for the New Architecture, we load the native entry point for
      // this app.
      DefaultNewArchitectureEntryPoint.load();
    }
    ReactNativeFlipper.initializeFlipper(this, getReactNativeHost().getReactInstanceManager());
    // FirebaseApp.initializeApp(this);
  }
}`

AndroidManifest.xml

`<manifest xmlns:android="http://schemas.android.com/apk/res/android"
  package="com.digianafresh" xmlns:tools="http://schemas.android.com/tools">

  <uses-permission android:name="android.permission.INTERNET" />
  <uses-permission android:name="android.permission.WAKE_LOCK" />
  <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
  <uses-permission android:name="android.permission.VIBRATE" />
  <uses-permission android:name="${applicationId}.permission.C2D_MESSAGE" />
  <uses-permission android:name="com.google.android.c2dm.permission.RECEIVE" />

  <permission
    android:name="${applicationId}.permission.C2D_MESSAGE"
    android:protectionLevel="signature" />

  <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED" />
  <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"
    tools:ignore="CoarseFineLocation" />
  <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
  <uses-permission
    android:name="com.google.android.gms.permission.AD_ID"
    tools:node="remove" />

  <application
    android:name=".MainApplication"
    android:label="@string/app_name"
    android:icon="@mipmap/ic_launcher"
    android:roundIcon="@mipmap/ic_launcher_round"
    android:theme="@style/AppTheme"
    tools:node="replace">

    <meta-data
      android:name="com.google.firebase.messaging.notification_color"
      android:resource="@color/white" />
    
    
    
    <activity
      android:name=".MainActivity"
      android:label="@string/app_name"
      android:theme="@style/SplashTheme"
      android:icon="@mipmap/ic_launcher"
      android:roundIcon="@mipmap/ic_launcher_round"
      android:configChanges="keyboard|keyboardHidden|orientation|screenLayout|screenSize|smallestScreenSize|uiMode"
      android:launchMode="singleTask"
      android:windowSoftInputMode="adjustResize"
      android:exported="true">
      <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <action android:name="com.google.firebase.MESSAGING_EVENT" />
        <category android:name="android.intent.category.LAUNCHER" />
      </intent-filter>
    </activity>
    <activity android:name="com.payu.ui.view.activities.CheckoutActivity"
      android:exported="false" />
    <meta-data android:name="com.google.android.geo.API_KEY"
      android:value="xxxx" />
    <meta-data android:name="com.google.android.gms.version" android:value="12451000" />
    <meta-data
      android:name="google_analytics_adid_collection_enabled"
      android:value="false" />
    

    
    <service android:name="com.google.firebase.components.ComponentDiscoveryService"
      android:directBootAware="true" android:exported="false">
      <meta-data
        android:name="com.google.firebase.components:com.google.firebase.messaging.FirebaseMessagingRegistrar"
        android:value="com.google.firebase.components.ComponentRegistrar" />

      <meta-data
        android:name="com.google.firebase.components:com.google.firebase.installations.FirebaseInstallationsRegistrar"
        android:value="com.google.firebase.components.ComponentRegistrar" />
    </service>
    
    <receiver
      android:name="com.google.firebase.messaging.FirebaseMessagingReceiver"
      android:exported="true"
      android:permission="com.google.android.c2dm.permission.SEND">
      <intent-filter>
        <action android:name="com.google.android.c2dm.intent.RECEIVE" />
        <category android:name="${applicationId}" />
      </intent-filter>
    </receiver>
    
    <service
      android:name="com.google.firebase.messaging.FirebaseMessagingService"
      android:exported="true">
      <intent-filter>
        <action android:name="com.google.firebase.MESSAGING_EVENT" />
      </intent-filter>
    </service>
    
  </application>

</manifest>`

**Comment by mikehardy** [comment]
The formatting here is very difficult to read. triple-backticks and a language-specifier is how you do it

There is never a need to do this and it may cause problems:

> firebaseMessagingVersion:"21.1.0"

also

```groovy
implementation platform('com.google.firebase:firebase-bom:32.7.0')
implementation('com.google.firebase:firebase-analytics')
implementation('com.google.firebase:firebase-messaging')
```

I'm not sure why are you directly depending on those when the module files from this package do it for you, nor am I sure why you specify a BOM here (using our *use at your own risk* override mechanism https://rnfirebase.io/#overriding-native-sdk-versions ) but I'll assume you have your reasons. If things break when you do that, you need to not do that though - I would suggest removing all four of the lines I've just mentioned here

Finally, you say "application getting exit" but you have not posted any stack trace indicating what the crash was and why it is from this project. You should use `adb logcat` to watch the device logs and include a crash stack trace if you are reporting a crash - then we can help

**Comment by ashish-hexalitics** [comment]
i removed these three line 
firebaseMessagingVersion:"21.1.0"
implementation platform('com.google.firebase:firebase-bom:32.7.0')
implementation('com.google.firebase:firebase-analytics')
i Didn't remove implementation('com.google.firebase:firebase-messaging') because i need this

and i attached debugger now i am getting this error
error is in blow screenshot

![image](https://github.com/invertase/react-native-firebase/assets/141706994/37e647fe-a9ba-47f0-9631-ee5c84aa3af7)

**Reply by mikehardy** [reply]
Your application is building and running correctly as far as this repository is concerned.
Now you are having project-specific programming errors, and we do not offer support for project-specific programming - developers must fix their own project code

If there is a bug in the compilation of the modules or one of the APIs is not meeting its contract, then that is something we support
